src/com/xilinx/rapidwright/rwroute/GlobalSignalRouting.java

```diff
@@ -0,0 +1,854 @@
+/*
+ * 
+ * Copyright (c) 2021 Ghent University. 
+ * All rights reserved.
+ *
+ * Author: Yun Zhou, Ghent University.
+ *
+ * This file is part of RapidWright. 
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * 
+ */
+
+package com.xilinx.rapidwright.rwroute;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+import com.xilinx.rapidwright.design.Design;
+import com.xilinx.rapidwright.design.Net;
+import com.xilinx.rapidwright.design.NetType;
+import com.xilinx.rapidwright.design.SiteInst;
+import com.xilinx.rapidwright.design.SitePinInst;
+import com.xilinx.rapidwright.device.ClockRegion;
+import com.xilinx.rapidwright.device.Device;
+import com.xilinx.rapidwright.device.IntentCode;
+import com.xilinx.rapidwright.device.Node;
+import com.xilinx.rapidwright.device.PIP;
+import com.xilinx.rapidwright.device.Site;
+import com.xilinx.rapidwright.device.Tile;
+import com.xilinx.rapidwright.device.Wire;
+import com.xilinx.rapidwright.placer.blockplacer.Point;
+import com.xilinx.rapidwright.placer.blockplacer.SmallestEnclosingCircle;
+import com.xilinx.rapidwright.router.RouteNode;
+import com.xilinx.rapidwright.router.RouteThruHelper;
+import com.xilinx.rapidwright.router.UltraScaleClockRouting;
+import com.xilinx.rapidwright.timing.ClkRouteTiming;
+import com.xilinx.rapidwright.util.Pair;
+
+/**
+ * A collection of methods for routing global signals, i.e. GLOBAL_CLOCK, VCC and GND.
+ * Adapted from RapidWright APIs.
+ */
+public class GlobalSignalRouting {
+	private static Design design;
+	private static RouteThruHelper routeThruHelper;
+	
+	static boolean clkDebug = false;
+	static boolean debugPrintClkPIPs = false;
+	
+	static Map<String, List<String>> crRoutes;
+	static Map<Pair<String, String>, List<Short>> bufceRowTaps;
+	static Device device;
```

Comment:
There is a reference to a `Device` here, but there are many methods that take as a parameter a `Device`.  This is confusing.  Again, we probably need to discuss this class -- if it should be static or not.  There are ways of getting a `Device` object (from a `Tile` for example) without having to require the user to provide it.  We can also add helper methods to avoid cumbersome APIs.